DOC-12950 CBL 3.2.2 doc

ElliotFrancisHunter opened the pull request:
## PR for Couchbase Lite version 3.2.2. 

### Preview Credentials: https://couchbasecloud.atlassian.net/wiki/spaces/DOCS/pages/1971224949/Docs+Team+demo+site+passwords

### Release Notes Reference:
https://confluence.issues.couchbase.com/wiki/spaces/cbeng/pages/2770305028/CBL3.2.2ReleaseNotes

## Changed Pages
-------------------

- What's new: https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/cbl-whatsnew.html#release-3-2-2-feburary-2025

- Release Notes (Android): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/android/releasenotes.html
- Release Notes (C): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/c/releasenotes.html
- Release Notes (.NET): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/csharp/releasenotes.html
- Release Notes (Java): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/java/releasenotes.html
- Release Notes (Objective-C): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/objc/releasenotes.html
- Release Notes (Swift): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/swift/releasenotes.html

- New Logging API (Android): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/android/new-logging-api.html
- New Logging API (C): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/c/new-logging-api.html
- New Logging API (.NET): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/csharp/new-logging-api.html
- New Logging API (Java): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/java/new-logging-api.html
- New Logging API (Objective-C): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/objc/new-logging-api.html
- New Logging API (Swift): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/swift/new-logging-api.html

- Partial Index (Android): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/android/indexing.html#partial-index
- Partial Index (C): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/c/indexing.html#partial-index
- Partial Index (.NET): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/csharp/indexing.html#partial-index
- Partial Index (Java): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/java/indexing.html#partial-index
- Partial Index (Objective-C): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/objc/indexing.html#partial-index
- Partial Index (Swift): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/swift/indexing.html#partial-index

- Old Logging API deprecation Notice (Android): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/android/troubleshooting-logs.html#introduction
- Old Logging API deprecation Notice (.NET): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/csharp/troubleshooting-logs.html#introduction
- Old Logging API deprecation Notice (Java): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/java/troubleshooting-logs.html#introduction
- Old Logging API deprecation Notice (Objective-C): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/objc/troubleshooting-logs.html#introduction
- Old Logging API deprecation Notice (Swift): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/swift/troubleshooting-logs.html#introduction

----------------------
## Pages still needing changes/updates
-----------------------

....

-----------------------

ElliotFrancisHunter committed:
DOC-12950: Update platform versions

ElliotFrancisHunter committed:
DOC-12950: 3.2.2 release note implementation

ElliotFrancisHunter committed:
DOC-12950: Partial Index copy

V1 before the addition of snippets to fully flesh out the content

ElliotFrancisHunter committed:
DOC-12950: Initial logging api prose

Initial prose before edits to accompany the various platform snippets

Co-Authored-By: G. Blake Meike <bmeike@users.noreply.github.com>

ElliotFrancisHunter committed:
Merge branch 'release/3.2' into DOC-12950-3-2-2-doc

ElliotFrancisHunter committed:
DOC-12950: 3.2.2 What's New copy

ElliotFrancisHunter committed:
DOC-12950: Implement new logging API page

Change list:
-- Common partial for the new logging API copy (still need to add platform snippets)
-- Remove new logging API copy from upgrade doc
-- Add deprecation notice to existing logging doc and signpost to the new doc
-- Create platform pages that include content from the common partial doc file

ElliotFrancisHunter committed:
DOC-12950: V1 partial index doc

Updated common partial to include partial index copy and snippets.
Still missing .NET snippets

RayOffiah commented on modules/ROOT/pages/_partials/commons/common-new-logging-api.adoc:
```diff
@@ -0,0 +1,84 @@
+
+A lot of the new logging API will seem very familiar: many of the concepts are
+the same.  There are still three places that log messages can be sent: the log
+files, the console, and custom code and Couchbase still logs its messages in a
```
```suggestion
files, the console, and custom code. Couchbase still logs its messages in a
```

bmeike replied:
This change is grammatically incorrect.  Adding the "and" makes it seem that there are four log sinks.  There are three: file, console and custom.  The next sentence describes the first of these, the file log sink.  Perhaps it should begin a new paragraph?

ElliotFrancisHunter replied:
Resolving as stale

RayOffiah commented on modules/ROOT/pages/_partials/commons/common-new-logging-api.adoc:
```diff
@@ -0,0 +1,84 @@
+A lot of the new logging API will seem very familiar: many of the concepts are
+the same.  There are still three places that log messages can be sent: the log
+files, the console, and custom code and Couchbase still logs its messages in a
+handfull of named "domains" and at common log levels: `debug` the most verbose,
```
```suggestion
handful of named "domains" and at common log levels: `debug` the most verbose,
```

ElliotFrancisHunter replied:
Done!

RayOffiah commented on modules/ROOT/pages/_partials/commons/common-indexing.adoc:
```diff
@@ -109,6 +109,215 @@ include::{root-partials}block_tabbed_code_example.adoc[]
+[source, sqlpp]
+----
+
+SELECT * FROM tab2 WHERE b=456;  -- cannot use partial index
```
Quick question:
This cannot use a partial index; is that because  the query isn't AND-connected?

ElliotFrancisHunter replied:
Pretty much! The index created in the snippet below for reference:

`CREATE INDEX ex2 ON tab2(b,c) WHERE c IS NOT NULL;`

The key thing here is WHERE c IS NOT NULL. The second query is related to the IMPORTANT admonition. It requires satisfaction of the first rule which is fails. The WHERE clause is the created index `X` is not AND-connected and the WHERE clause in the query `W` is not OR-connected and contains none of the terms from the WHERE clause - as in it is missing reference to `c`

RayOffiah reviewed (dismissed):
Two small changes and a question to satisfy my curiosity.

bmeike reviewed (commented):
I do not think replacing the period with a conjunction is a good idea.

ElliotFrancisHunter committed:
DOC-12950: Fix tag name mismatch

ElliotFrancisHunter committed:
DOC-12950: New Logging API page v1

Copy edit for new logging api doc.
Adding snippets comparing old and new API methods.

ElliotFrancisHunter committed:
DOC-12950: Clarify admonition copy

ElliotFrancisHunter committed:
DOC-12950: Tidy copy

idulo commented on modules/ROOT/pages/cbl-whatsnew.adoc:
```diff
@@ -17,6 +17,82 @@ NOTE: Couchbase Lite 3.0 introduces some breaking changes. +
+
+* Removed polling overhead on custom loggers.
+
+* Simplified platform code and implementation.
```
this is more of a benefit for us, maybe for users: simplified API and reduced implementation complexity? let's put it last in the list as overall benefit

ElliotFrancisHunter replied:
Done!

bmeike commented:
Here: New Logging API (Android): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/android/new-logging-api.html
Only the last point is of any interest to the user.
The new API is, absolutely, not backwards compatible.  It is, internally, backwards compatible: the code we wrote will support the old API while it is deprecated but not removed.  For our users, it is completely incompatible.

bmeike commented:
Here: Old Logging API deprecation Notice (Android): https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/android/troubleshooting-logs.html#introduction
"setRotateCount from the old API is slightly different from setMaxKeptFiles. setRotateCount is the maximum size in bytes of a log file. The latter is the maximum number of log files that will exist at any time and is the count of rotated files plus one. "

This is not true.  `setRotateCount` has nothing to do with the size of the log file.  It is the number of rotated files. It is maxKeptFiles - 1

bmeike reviewed (dismissed):
There are several miss-communications here that should be corrected
I've commented them.  Holler if any of the comments are unclear.

ElliotFrancisHunter committed:
Merge branch 'release/3.2' into DOC-12950-3-2-2-doc

ElliotFrancisHunter committed:
DOC-12950: Update release date

Release moved to March

ElliotFrancisHunter committed:
DOC-12950: Add 3.2.2 to gs-downloads for C

ElliotFrancisHunter committed:
DOC-12950: Legacy logging API clarification

Change list:
-- Change title to reflect legacy nature of previous logging API
-- Rename tag for a duplicate code snippet

ElliotFrancisHunter committed:
DOC-12950: Address feedback

ElliotFrancisHunter committed:
DOC-12950: Typo fix

bmeike commented:
On this page:
https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/android/new-logging-api.html

== This:
"at common log levels: debug the most verbose, and error only for serious failures." should probably read:
 at common log levels: LogLevel.DEBUG the most verbose, and LogLevel.ERROR only for serious failures.
 
 
== This:
The new Logging API has the following benefits:
    Removed polling overhead on custom loggers.
    You can enque Log callbacks from LiteCore for asynchronous delivery.
    Log sinks are now thread safe, removing risk of inconsistent states during initialization.
    Simplified API and reduced implementation complexity.

Could be simplified.  Only the last two points are visible to client code.  The second point is probably incorrect....

== https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/android/new-logging-api.html#lbl-console-logsink
Perhaps those two logging levels should be the same?

== https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/android/new-logging-api.html#lbl-custom-logsink
I suggest removing this sentence entirely:
"Unless otherwise specified logs from all Couchbase domains are forwarded to the custom log sink."
The sentence below it is correct and this sentence appears to contradict it.

bmeike commented:
== https://preview.docs-test.couchbase.com/322Doc/couchbase-lite/current/java/new-logging-api.html#lbl-file-logsink
I think the note should read:
setRotateCount from the old API is slightly different from setMaxKeptFiles. The latter is the maximum number of log files that will exist at any time and is the count of rotated files (setRotateCount) plus one.

bmeike reviewed (approved):
I have a couple suggestions, but I don't see anything blocking

ElliotFrancisHunter committed:
DOC-12950: Feedback fixes

pasin commented on modules/ROOT/pages/_partials/commons/common-indexing.adoc:
```diff
@@ -109,6 +109,215 @@ include::{root-partials}block_tabbed_code_example.adoc[]
+[source, sqlpp]
+----
+
+CREATE INDEX ex2 ON tab2(b,c) WHERE c IS NOT NULL;
```
CBL doesn't support this create index syntax and we are using the API for creating the index. For simplicity, I think we can just say the following without using the actual code snippet:

```
For example, let the partial index be `c IS NOT NULL` when creating a partial index on collection named "col1" in the default scope.
```

simon-dew replied:
Done

pasin commented on modules/ROOT/pages/_partials/commons/common-indexing.adoc:
```diff
@@ -109,6 +109,215 @@ include::{root-partials}block_tabbed_code_example.adoc[]
+[source, sqlpp]
+----
+
+SELECT * FROM tab2 WHERE b=456 AND c<>0;  -- uses partial index
```
I suggested to use "col1" above as tab is more like a table name.

simon-dew replied:
Done

pasin commented on modules/ROOT/pages/_partials/commons/common-indexing.adoc:
```diff
@@ -109,6 +109,215 @@ include::{root-partials}block_tabbed_code_example.adoc[]
+[source, sqlpp]
+----
+
+SELECT * FROM tab2 WHERE b=456;  -- cannot use partial index
```
I suggested to use "col1" above as tab is more like a table name.

simon-dew replied:
Done

pasin commented on modules/ROOT/pages/_partials/commons/common-new-logging-api.adoc:
```diff
@@ -0,0 +1,634 @@
+
+The `FileLogSink`, the `ConsoleLogSink` and the `CustomLogSink` are all installed in an instance of `LogSinks`. 
+
+All three log sinks are initialized by default. 
```
It seems this could give an impression that all log sinks are enabled. I think it's more accurate to say that only ConsoleLogSink is enabled for all logging domains at the warning log level by default.

velicuvlad replied:
I second this. Moreover, I think it's "more correct" to maybe say that "all are to be installed in the LogSinks instance". There is only 1 instance of LogSinks and is available to the user right away.

simon-dew replied:
Reworded as recommended

pasin commented on modules/ROOT/pages/_partials/commons/common-new-logging-api.adoc:
```diff
@@ -0,0 +1,634 @@
+
+For example, you can stop logging to the FileLogSink like this:
+
+`LogSinks.get().setFile(null)`
```
This is very specific to Java / Android and as only console log sink is enabled by default, I think we could just say something like :

```
To enable a specific type of LogSink, create a LogSink object of that type, set its minimum log level and domains, and assign it to LogSinks. To disable a LogSink, set it to null or use a LogSink with LogLevel.NONE.
```

The code snippet for disabling the log sink should be shown here, we will need the code snippet for each platform. Maybe the current code snippet for setting each type of log sink is enough?

velicuvlad replied:
For the generated previews, this is the case for iOS as well - the Java specific snippet. More specifically, Swift should be `LogSinks.file = nil` instead. Similar for Obj-C `CBLLogSinks.file = nil`. And so, if we don't go the route to be adding additional snippets for invalidating sinks (disable logging), Pasin's suggestion is good/universal for all platforms.

simon-dew replied:
Reworded as suggested.

pasin commented on modules/ROOT/pages/_partials/commons/common-new-logging-api.adoc:
```diff
@@ -0,0 +1,634 @@
+
+Couchbase still logs its messages in a handful of named domains and at common log levels: `LogLevel.DEBUG` the most verbose, and `LogLevel.ERROR` only for serious failures.
+
+To get the instance of `LogsSinks`, use the method `LogSinks.get()`.
```
`LogSinks.get()` is Java/Android specific and it should be `LogSinks` not `LogsSinks`. 

Maybe, `Accessing LogSinks depends on the platform. In Java, use LogSinks.get() to obtain an instance. In Swift, Objective-C, .NET, and C, LogSinks provides only static methods, without a singleton instance.`

simon-dew replied:
Done

pasin commented on modules/ROOT/pages/_partials/commons/common-new-logging-api.adoc:
```diff
@@ -0,0 +1,634 @@
+=== Using a Custom Logger
+
+Installing a custom log sink with the new API is also streamlined: create an
+instance of your custom sink and use `LogSinks.get().setCustom` to install it.
```
Maybe `Use LogSinks.get().setCustom in Java or the corresponding static methods on other platforms.`

simon-dew replied:
Done

pasin commented on modules/ROOT/pages/_partials/commons/common-new-logging-api.adoc:
```diff
@@ -0,0 +1,634 @@
+change is that instead of _implementing_ the `Logger` interface, a custom log sink
+must _extend_ the `BaseLogSink` class.  
+
+The `BaseLogSink` class does not have a no-args constructor: you must specify at least the level at which Couchbase logs will be forwarded to the logger.  
```
Not sure what is the best way to describe this in the cross-platform way. The `BaseLogSink` described here only applies to Java/Android and .NET. 

In Swift and Objective-C, which lack abstract methods, a custom log sink implements LogSinkProtocol and is assigned to a CustomLogSink instance during creation.

Similarly, in C, the custom log sink implementation is a callback that is assigned to CBLCustomLogSink object during creation.

velicuvlad replied:
I think that for the custom log sink every platform should have its own rubric with a brief explanation and the specifics. There are snippets from old to new api just below that are platform specific as well and so it should suffice.

simon-dew replied:
Done

velicuvlad commented on modules/ROOT/pages/_partials/commons/common-new-logging-api.adoc:
```diff
@@ -0,0 +1,634 @@
+They were never meant as a general framework for logging. 
+
+IMPORTANT: With the new API, customer code can no longer log, directly, to any of the Couchbase log sinks.  
+The Console and File logsinks are final classes and do not publish methods that allow writing logs.
```
_logsinks are final classes_ 

Although, from the user perspective, both a final class and a struct cant be subclassed, this is not true for iOS. Swift LogSinks are structs and Obj-C doesn't have a final class concept (we "manually" prevent against subclassing.)

simon-dew replied:
Reworded "cannot be subclassed"

velicuvlad commented on modules/ROOT/pages/_partials/commons/common-new-logging-api.adoc:
```diff
@@ -0,0 +1,634 @@
+[source, swift]
+----
+
+include::swift:example$code_snippets/SampleCodeTest.swift[tags=new-file-logging]
```
I need to make an update to this - to add `maxSize`, such that is matching 1to1 with the old snippet.

simon-dew replied:
Okay, I see it's there

velicuvlad committed:
Swift - update new api snippet to match the old one

borrrden commented on modules/csharp/pages/new-logging-api.adoc:
```diff
@@ -0,0 +1,18 @@
+include::partial$_set_page_context_for_csharp.adoc[]
+
+// ROOT:partial$commons/common-
+include::{root-commons}new-logging-api.adoc[]
```
This partial contains Java style code snippets inline

simon-dew replied:
Okay, I _think_ these have all been dealt with by other fixes. Please take a look at [common-new-logging-api.adoc](https://github.com/couchbase/docs-couchbase-lite/blob/a9a73099daee390d27b788c3044f27d5d168a4f3/modules/ROOT/pages/_partials/commons/common-new-logging-api.adoc) and make sure I haven't missed anything.

borrrden replied:
Looks good to me

simon-dew replied:
Fantastic, thanks

borrrden commented on modules/ROOT/pages/_partials/commons/common-indexing.adoc:
```diff
@@ -109,6 +109,215 @@ include::{root-partials}block_tabbed_code_example.adoc[]
+
+[source, csharp, indent=0]
+----
+include::csharp:example$code_snippets/Program.cs[tags=array-index-config]
```
Why is C# only using a different tag than the others?

simon-dew replied:
Good spot, fixed

borrrden commented on modules/ROOT/pages/_partials/commons/common-indexing.adoc:
```diff
@@ -109,6 +109,215 @@ include::{root-partials}block_tabbed_code_example.adoc[]
+
+[source, csharp, indent=0]
+----
+include::csharp:example$code_snippets/Program.cs[tags=array-index-config]
```
Here C# is also using a different tag, this seems incorrect

simon-dew replied:
Good spot, fixed

borrrden reviewed (dismissed):
Also agree with what Pasin has said.

simon-dew committed:
Updates after review

simon-dew committed:
Minor formatting tweaks

pasin commented on modules/ROOT/pages/_partials/commons/common-indexing.adoc:
```diff
@@ -109,14 +109,201 @@ include::{root-partials}block_tabbed_code_example.adoc[]
+
+For example, let the partial index be `c IS NOT NULL` when creating a partial index on collection named `col1` in the default scope.
+
+Then any query that uses operators `=`, `<`, `>`, `\<=`, `>=`, `<>`, `IN`, `LIKE`, or `GLOB` on column `c` would be usable with the partial index, because those comparison operators are only true if `c` is not NULL. 
```
Just noticed this, we don't support GLOB and BETWEEN is missing. So basically, changing BLOB to BETWEEN.

simon-dew replied:
Done

pasin reviewed (approved):
Approved with one small change request.

simon-dew committed:
Minor comparison operator fix

idulo commented:
@simon-dew on the Partial Index support, we have a section: Partial Full-Text Index, so we call it Full Text, but in practice this is related to the Full Text Search (FTS) feature and it should be with consistent naming. Also we should deep link to the explanation around FTS and the match() function -> e.g. https://docs.couchbase.com/couchbase-lite/current/c/fts.html#use-index as user might not have context for what what we currently call Full Text is and how it can be used.

simon-dew commented:
> @simon-dew think we should pull the benefits of the new Logging API earlier in the section and communicate it is easy to migrate, encourage earlier migration.

Done

simon-dew committed:
Boost benefits of new logging API;
clarify partial full-text indexes
